### ISSUE TITLE
Model evaluation is wrong when integrate setting changed; a cache issue?

### ISSUE THREAD

**Issue by DougBurke** [sherpa/sherpa#958]
This got created twice - #958 and #959. I'm closing #959 but there was some useful commentary I added to it.

# Details

With the following program

```
import numpy as np
from sherpa.models.basic import Polynom1D

m1 = Polynom1D()
m2 = Polynom1D()

m1.integrate = False
m2.integrate = True

xgrid = np.arange(2, 10, 2)
xlo, xhi = xgrid[:-1], xgrid[1:]

m1.c0 = 5
m2.c0 = 5
m1.c1 = 1.5
m2.c1 = 1.5

print("m1(x)")
print(m1(xgrid))

print("m2(x)")
print(m2(xgrid))

print("m1(xlo, xhi)")
print(m1(xlo, xhi))

print("m2(xlo, xhi)")
print(m2(xlo, xhi))

print("Changing m1.integrate")
m1.integrate = True

print("m1(xlo, xhi)")
print(m1(xlo, xhi))
```

I would expect the last line to display the same output as m2(xlo, xhi), since the parameters are the same as is the integrate setting. It isn't - it is in fact returning the integrate=False output for m1,  which I *assume* is because of some bizarre cache error.

I have noticed occasional strange results which I believe due to be related to this behavior.

The output I get is

```
m1(x)
[ 8. 11. 14. 17.]
m2(x)
[ 8. 11. 14. 17.]
m1(xlo, xhi)
[ 8. 11. 14.]
m2(xlo, xhi)
[19. 25. 31.]
Changing m1.integrate
m1(xlo, xhi)
[ 8. 11. 14.]
```

**Comment by dtnguyen2** [comment]
Add a print statement to see if the variable ```integrate``` is passed into the ```modelCacher1d``` or not:

```
(sherpa) [dtn@devel12 sherpa]$ git diff
diff --git a/sherpa/models/model.py b/sherpa/models/model.py
index 70fbe89f..daf93aec 100644
--- a/sherpa/models/model.py
+++ b/sherpa/models/model.py
@@ -55,6 +55,7 @@ def modelCacher1d(func):
         digest = ''
         if use_caching:
 
+            print('modelCacher1d: integrate =', kwargs.get('integrate', False))
             data = [numpy.array(pars).tobytes(),
                     boolean_to_byte(kwargs.get('integrate', False)),
                     numpy.asarray(xlo).tobytes()]
```

It appears that the ```integrate``` was never passed  into ```modelCacher1d``` via ```kwargs```:
```
m1(x)
modelCacher1d: integrate = False
[ 8. 11. 14. 17.]
m2(x)
modelCacher1d: integrate = False
[ 8. 11. 14. 17.]
m1(xlo, xhi)
modelCacher1d: integrate = False
[ 8. 11. 14.]
m2(xlo, xhi)
modelCacher1d: integrate = False
[19. 25. 31.]
Changing m1.integrate
m1(xlo, xhi)
modelCacher1d: integrate = False
[ 8. 11. 14.]
```

I suspect ```modelCacher1d``` will behave correctly if (when?) ```kwargs``` contains ```integrate```

**Comment by DougBurke** [comment]
@dtnguyen2 - I had added https://github.com/sherpa/sherpa/issues/959#issuecomment-703852425 to #959 just before you sent your comment, but the fact these were duplicated comments meant I didn't see it.

My fix in #960 doesn't use the kwarg integrate setting because we have code like the following in `sherpa.models.basic`:

    @modelCacher1d
    def calc(..., **kwargs):
        kwargs['integrate'] = self.integrate
        return some_method(..., **kwargs)

and the problem is that the kwargs[integrate] setting is too late here - it would have to gave been used before calling calc, since it is calc that gets wrapped by the modelCacher1d decorator, and not some_method.

Instead the code now just uses the integrate setting of the model that is being wrapped, with a fall through to kwarg if there is no integrate setting, and a fall through to False if the kwarg isn't set. 

My current belief is that the integrate kwarg setting should be removed from all the models, as it's not clear it provides any benefit, but it's a complicated-enough setup that this is definitely up for debate.
